lang/src/main/kotlin/org/partiql/lang/eval/ExprValue.kt

```diff
@@ -56,3 +78,114 @@ interface ExprValue : Iterable<ExprValue>, Faceted {
+            }
+        }
+    }
+}
+
+fun IonValue.toExprValue(): ExprValue {
+    val valueFactory = ExprValueFactory.standard(system)
+    return when {
+        isNullValue && hasTypeAnnotation(MISSING_ANNOTATION) -> valueFactory.missingValue // MISSING
+        isNullValue -> NullExprValue(system, type) // NULL
+        this is IonBool -> valueFactory.newBoolean(booleanValue()) // BOOL
+        this is IonInt -> valueFactory.newInt(longValue()) // INT
+        this is IonFloat -> valueFactory.newFloat(doubleValue()) // FLOAT
+        this is IonDecimal -> valueFactory.newDecimal(decimalValue()) // DECIMAL
+        this is IonTimestamp && hasTypeAnnotation(DATE_ANNOTATION) -> {
+            val timestampValue = timestampValue()
+            valueFactory.newDate(timestampValue.year, timestampValue.month, timestampValue.day)
+        } // DATE
+        this is IonTimestamp -> valueFactory.newTimestamp(timestampValue()) // TIMESTAMP
+        this is IonStruct && hasTypeAnnotation(TIME_ANNOTATION) -> {
+            val hourValue = (this["hour"] as IonInt).intValue()
+            val minuteValue = (this["minute"] as IonInt).intValue()
+            val secondInDecimal = (this["second"] as IonDecimal).decimalValue()
+            val secondValue = secondInDecimal.toInt()
+            val nanoValue = secondInDecimal.remainder(BigDecimal.ONE).multiply(NANOS_PER_SECOND.toBigDecimal()).toInt()
+            val timeZoneHourValue = (this["timezone_hour"] as IonInt).intValue()
+            val timeZoneMinuteValue = (this["timezone_minute"] as IonInt).intValue()
+            valueFactory.newTime(Time.of(hourValue, minuteValue, secondValue, nanoValue, secondInDecimal.scale(), timeZoneHourValue * 60 + timeZoneMinuteValue))
+        } // TIME
+        this is IonSymbol -> valueFactory.newSymbol(stringValue()) // SYMBOL
+        this is IonString -> valueFactory.newString(stringValue()) // STRING
+        this is IonClob -> valueFactory.newClob(bytesValue()) // CLOB
+        this is IonBlob -> valueFactory.newBlob(bytesValue()) // BLOB
+        this is IonList && hasTypeAnnotation(BAG_ANNOTATION) -> valueFactory.newBag(map { it.toExprValue() }) // BAG
+        this is IonList -> valueFactory.newList(map { it.toExprValue() }) // LIST
+        this is IonSexp -> valueFactory.newSexp(map { it.toExprValue() }) // SEXP
+        this is IonStruct -> IonStructExprValue(valueFactory, this) // STRUCT
+        else -> error("Unrecognized IonValue to transform to ExprValue: $this")
+    }
+}
+
+private class IonStructExprValue(
+    valueFactory: ExprValueFactory,
+    private val ionStruct: IonStruct
+) : StructExprValue(
+    valueFactory.ion,
+    StructOrdering.UNORDERED,
+    ionStruct.asSequence().map {
+        it.toExprValue().namedValue(valueFactory.newString(it.fieldName))
+    }
+) {
+    override val bindings: Bindings<ExprValue> =
+        IonStructBindings(valueFactory, ionStruct)
+
+    override val ionValue: IonValue
+        get() = ionStruct
+}
```

Comment:
What was this created for? Just asking out of curiosity. It's a little difficult for me to pick out since this is a large PR. Is this for an explicit struct instead of partiql bindings?

Reply:
See https://github.com/partiql/partiql-lang-kotlin/issues/908

Reply:
This will be removed eventually. Ion structs will be the same as PartiQL structs.